src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -234,15 +226,132 @@ static void CheckForExperimentStatus(ITextView textView, in CompletionOptions op
+                AsyncCompletionLogger.LogImportCompletionGetContext(isBlocking: true, delayed: false);
+                return context;
+            }
+            else
             {
-                return AsyncCompletionData.CompletionContext.Empty;
+                // OK, expand item is enabled but we shouldn't block completion on its results.
+                // Kick off expand item calculation first in background.
+                Stopwatch stopwatch = new();
+                var expandedItemsTask = Task.Run(async () =>
+                {
+                    var result = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                      options with { ExpandedCompletionBehavior = ExpandedCompletionMode.ExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+
+                    // Record how long it takes for the background task to complete *after* core providers returned.
+                    // If telemetry shows that a short wait is all it takes for ExpandedItemsTask to complete in
+                    // majority of the sessions, then we might consider doing that instead of return immediately.
+                    var elapsedMs = (int)stopwatch.ElapsedMilliseconds;
+                    if (elapsedMs > 0)
+                        AsyncCompletionLogger.LogAdditionalTicksToCompleteDelayedImportCompletionDataPoint(elapsedMs);
+
+                    return result;
+                }, cancellationToken);
+
+                // Now trigger and wait for core providers to return;
+                var (nonExpandedContext, nonExpandedCompletionList) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                        options with { ExpandedCompletionBehavior = ExpandedCompletionMode.NonExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+                AddPropertiesToSession(session, nonExpandedCompletionList, triggerLocation);
+
+                if (expandedItemsTask.IsCompleted)
+                {
+                    // the task of expanded item is completed, get the result and combine it with result of non-expanded items.
+                    var (expandedContext, expandedCompletionList) = await expandedItemsTask.ConfigureAwait(false);
+                    AddPropertiesToSession(session, expandedCompletionList, triggerLocation);
+                    AsyncCompletionLogger.LogImportCompletionGetContext(isBlocking: false, delayed: false);
+
+                    return CombineCompletionContext(nonExpandedContext, expandedContext);
+                }
+                else
+                {
+                    // Expanded item task still running. Save it to the session and return non-expanded items immediately.
+                    // Also start the stopwatch since we'd like to know how long it takes for the expand task to finish
+                    // after core providers completed (instead of how long it takes end-to-end).
+                    stopwatch.Start();
```

Comment:
This doesn't really affect the user experience, but this can race with the read in `expandedItemsTask` right? 
 Because if the stopwatch is started after the `if elapsedMs > 0` then we record a delayed completion event, but don't get the number of ticks delayed in telemetry.

Would it be easier to just record both the normal completion task time and the expanded completion task time, then when logging telemetry just wait for both to finish and calculate the diff?

Reply:
Yes, there'd be race, I ignored that since I didn't think we need high level of accuracy for this. Now that you mentioned it, I think it'd probably better to log it regardless of whether `stopwatch` is started (remove the >0 check), which would give us a better idea as of how much longer import completion task takes (even if it's finished almost in time but ended up not included.) 

The issue with recording time for both tasks is we trigger the import completion task first, and we are more interested in the additional time needed after the regular task completed. Maybe it doesn't matter, I'm not sure

Reply:
Ah - so in the cases where it finishes before you just record a 0 additional time?  That seems to make sense and would be straightforward.

Reply:
Yup, and we have another property to tell us how many time it's delayed